### PR TITLE
refactor(callback) : email

### PR DESCRIPTION
**Pull request by jeantanzj**
I removed support for SNS events since the SES topics will only be subscribed to the new http endpoints, and not the lambda. 

### SES
- Create a subscription to the staging ses topic in eu-central-1 like so, to your local tunnel. If you need to confirm the subscription, add this line to get the subscribe url: https://github.com/opengovsg/postmangovsg/pull/684/commits/d67a3e44955f05e0a61a69f11d61d93ed25b39f5
Screenshot 2020-09-03 at 12 19 38 AM

- Use the  eu-central-1 credentials to send out messages 

### SendGrid
As our SendGrid account is currently inactive, it's cumbersome to test.
This is the content of a callback for a delivered message:
```
[{"email":"postman@pf.postman.gov.sg","environment":"development","event":"delivered","ip":"167.89.12.138","message_id":"10","response":"250 2.0.0 Ok: queued as BE79A4383C","sg_event_id":"ZGVsaXZlcmVkLTAtMTcxMTkxMjMtQl9MalhPcEtTb0MtVk5WWkFrODRXQS0w","sg_message_id":"B_LjXOpKSoC-VNVZAk84WA.filterdrecv-p3las1-7754f7d4cc-cghdw-18-5F046149-20.0","smtp-id":"<0bd065b4-de88-f6da-2b05-fdb66058f3e3@mail2.postman.gov.sg>","timestamp":1594122578,"tls":1}]
```
This is the callback for a bounced message:
```
[{"email":"bounce2@zyx.abc.sg","environment":"development","event":"bounce","ip":"167.89.12.138","message_id":"10","reason":"550 5.1.1 The email account that you tried to reach does not exist. Please try double-checking the recipient's email address for typos or unnecessary spaces. Learn more at https://support.google.com/mail/?p=NoSuchUser lv19si13071710ejb.241 - gsmtp","sg_event_id":"Ym91bmNlLTAtMTcxMTkxMjMtTGIwR2dXamVUeDJyS1J1WktRbDRZdy0w","sg_message_id":"Lb0GgWjeTx2rKRuZKQl4Yw.filterdrecv-p3iad2-5b55dcd864-kvwkt-20-5F03B661-4.0","smtp-id":"<f78991f8-92e6-2ac6-537f-61b882e8c3ab@mail2.postman.gov.sg>","status":"5.1.1","timestamp":1594078818,"tls":1,"type":"bounce"}]
```

- [ ] An email that's successfully sent should have its status updated to SUCCESS
- [ ] An email that bounced should have its status updated to INVALID_RECIPIENT, and error_code updated to either Soft bounce or Hard bounce. 
- [ ] An email that was reported as spam should have its status updated to INVALID_RECIPIENT, and error_code updated to 'abuse'. 
- [ ] If the thresholds for halting are met, the column 'halted' in the campaigns table should be updated to 'True'.

### CODE REVIEW

File: backend/.env-example [hunk 1]
```diff
@@ -10,7 +10,14 @@ SECRET_MANAGER_SALT=$2b$10$iDQWc/ZxTnAOMdx8022VWe
 JWT_SECRET=secret
 API_KEY_SALT_V1=$2b$10$5CuueB/CRZq8/JgmkzAK1e
 UNSUBSCRIBE_HMAC_ALGO_V1=sha256
-UNSUBSCRIBE_HMAC_KEY_V1=secret
+
```
Comment (lshaowei18) [inline backend/.env-example:13]:
```suggestion
UNSUBSCRIBE_HMAC_KEY_V1=secret
```

File: backend/src/email/utils/callback/update-status.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+ *  @param timestamp ISO string from notification timestamp
+ */
+export const updateComplaintStatus = async (
+  metadata: ComplaintMetadata
+): Promise<void> => {
+  const errorCode = metadata.complaintType
+  const recipients = metadata.to
+
+  if (errorCode && recipients) {
+    await Promise.all(recipients.map(addToBlacklist))
```
Comment (lshaowei18) [inline backend/src/email/utils/callback/update-status.ts:60]: Sorry I can't recall whats the discussion around this. With the unsubscribe digest, is it still necessary to add the complaints to blacklist?

**Comment by lshaowei18** [comment]
- [ ] An email that's successfully sent should have its status updated to SUCCESS
- [ ] An email that bounced should have its status updated to INVALID_RECIPIENT, and error_code updated to either Soft bounce or Hard bounce. 
- [ ] An email that was reported as spam should have its status updated to INVALID_RECIPIENT, and error_code updated to 'abuse'. 
- [ ] If the thresholds for halting are met, the column 'halted' in the campaigns table should be updated to 'True'.

File: backend/.env-example (hunk 1 above)
Comment (jeantanzj) [reply backend/.env-example:13]: Sorry what is this change?

File: backend/src/email/utils/callback/update-status.ts (hunk 2 above)
Comment (jeantanzj) [reply backend/src/email/utils/callback/update-status.ts:60]: You're right in a way that users might be more inclined to click unsubscribe than to mark as spam. But if the user sends an unsubscribe request, and the agency user doesn't respect it and remove that email address from the recipients list, they're still going to receive messages from Postman, which they'll continue marking as spam and affect our reputation.

**Comment by jeantanzj** [comment]
> SES
> 
> * [x]  An email that's successfully sent should have its status updated to SUCCESS
> * [x]  An email that bounced should have its status updated to INVALID_RECIPIENT, and error_code updated to either Soft bounce or Hard bounce.
> * [ ]  An email that was reported as spam should have its status updated to INVALID_RECIPIENT, and error_code updated to 'abuse'.
> * [x]  If the thresholds for halting are met, the column 'halted' in the campaigns table should be updated to 'True'.
> 
> I reported an email as spam, but I didn't receive the callback/wasn't sure if gmail did report.

https://docs.aws.amazon.com/ses/latest/DeveloperGuide/send-email-simulator.html
For the spam test, use `complaint@simulator.amazonses.com` . Remember to delete that address from the blacklist table first if it exists, otherwise the message wont get sent.

File: backend/src/email/middlewares/email-callback.middleware.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
+import { Request, Response, NextFunction } from 'express'
+import { EmailCallbackService } from '@email/services'
+
+const isAuthenticated = (
+  req: Request,
+  res: Response,
+  next: NextFunction
+): Response | void => {
+  const authHeader = req.get('authorization')
```
Comment (jeantanzj) [inline backend/src/email/middlewares/email-callback.middleware.ts:9]: note to self: we need to enable subscription confirmation to this endpoint first, then do a second release to get rid of the subscription confirmation code.

File: backend/.env-example (hunk 1 above)
Comment (lshaowei18) [reply backend/.env-example:13]: Hmmm do we still need `UNSUBSCRIBE_HMAC_KEY_V1`?

**Review by lshaowei18 — APPROVED** [review]

**Comment by lshaowei18** [comment]
Didn't managed to test sendgrid, SES lgtm!

File: backend/src/email/utils/callback/parsers/sendgrid.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+type SendgridEvent = SendgridRecord[]
+
+type SendgridRecord = {
+  // the unique_args that we passed
+  message_id: string
+  // standard args
+  email: string
+  event: string
+  type: string
+  'smtp-id': string
```
Comment (lshaowei18) [inline backend/src/email/utils/callback/parsers/sendgrid.ts:28]: Sorry why is the property in quotes? `'smtp-id'`

Comment (jeantanzj) [reply backend/src/email/utils/callback/parsers/sendgrid.ts:28]: because it's not just alphanumeric? if it were `smtpId` then you wouldn't need quotes.

Comment (lshaowei18) [reply backend/src/email/utils/callback/parsers/sendgrid.ts:28]: Ah thanks I didn't know that

File: backend/.env-example (hunk 1 above)
Comment (jeantanzj) [reply backend/.env-example:13]: oh ok thanks for the catch

**Pull request merged**